teachertool/src/services/makecodeEditorService.ts

```diff
@@ -0,0 +1,52 @@
+interface PendingMessage {
+    original: pxt.editor.EditorMessageRequest;
+    handler: (response: any) => void;
+}
+
+let makecodeEditorRef: HTMLIFrameElement | undefined;
+const messageQueue: pxt.editor.EditorMessageRequest[] = [];
+let nextId: number = 0;
+let pendingMessages: {[index: string]: PendingMessage} = {};
+
+const onMessageReceived = (event: MessageEvent) => {
+    const data = event.data as pxt.editor.EditorMessageRequest;
+    if (data.type === "pxteditor" && data.id && pendingMessages[data.id]) {
+        const pending = pendingMessages[data.id];
+        pending.handler(data);
+        delete pendingMessages[data.id];
+        return;
+    }
+}
+
+export const sendMessageAsync = (message?: any) => {
```

Comment:
We shouldn't expose `sendMessageAsync`. It is too low level. Instead, create methods for specific messages, like `setHighContrast(on: boolean)`. For methods that expect a reply, return a Promise (strongly typed to the reply message). The obvious benefit to this approach is that it keeps application code very clean. Another is that within makecodeEditorService, you can decide which messages need to be kept in PendingMessages. Only the messages that expect a response need to be tracked this way (though you would need another queue to pend messages sent before the iframe is ready, but I think you need this anyway).

Reply:
+1 this will keep all of the "nitty gritty" details (communicating with the iframe, what messages to send, etc...) contained within this class, then the rest of the code doesn't have to worry about it.